### PR TITLE
Deferrable mode for Custom Training Job operators

### PR DESCRIPTION
**Pull request by e-galan**
The PR adds the deferrable mode feature for the following Vertex AI operators:
- CreateCustomContainerTrainingJobOperator
- CreateCustomPythonPackageTrainingJobOperator
- CreateCustomTrainingJobOperator

Other changes include:
- Fix for `VertexAIModelLink`. The old link url did not work anymore, I replaced it with a working one.
- New unit tests
- Update example dags

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by VladaZakharova** [comment]
Hi there! 
@eladkal  @Lee-W @potiuk 
Can you please check the PR here? :)

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py [hunk 1]
```diff
@@ -1754,79 +1774,1181 @@ def create_custom_training_job(
...
         project_id: str,
         region: str,
-        pipeline_job: str,
-        retry: Retry | _MethodDefault = DEFAULT,
-        timeout: float | None = None,
-        metadata: Sequence[tuple[str, str]] = (),
-    ) -> Operation:
+        display_name: str,
+        container_uri: str,
+        command: Sequence[str] = [],
```
Comment (Lee-W) [inline airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py]: It's not suggested to use mutable objects as default values.

Comment (Lee-W) [reply airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py]: We probably need to set it to None and initialize it in this method instead

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py [hunk 2]
```diff
@@ -1754,79 +1774,1181 @@ def create_custom_training_job(
...
+        model_version_aliases: list[str] | None = None,
+        model_version_description: str | None = None,
+        labels: dict[str, str] | None = None,
+        training_encryption_spec_key_name: str | None = None,
+        model_encryption_spec_key_name: str | None = None,
+        staging_bucket: str | None = None,
+        # RUN
+        dataset: None
+        | (
+            datasets.ImageDataset | datasets.TabularDataset | datasets.TextDataset | datasets.VideoDataset
```
Comment (Lee-W) [inline airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py:1807]: Should we make a type alias to make it more readable?

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py [hunk 3]
```diff
@@ -2178,3 +3300,239 @@ def list_custom_jobs(
...
+        :param timeout: The timeout for this request.
+        :param metadata: Strings which should be sent along with the request as metadata.
+        """
+        client = self.get_pipeline_service_client(region)
+        name = client.pipeline_job_path(project_id, region, pipeline_job)
+
+        result = client.delete_pipeline_job(
+            request={
+                "name": name,
+            },
```
Comment (Lee-W) [inline airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py]: comma was not added elsewhere. let's stick with one convention
```suggestion
            request={"name": name},
```

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py [hunk 4]
```diff
@@ -2178,3 +3300,239 @@ def list_custom_jobs(
...
+            except Exception as ex:
+                self.log.exception("Exception occurred while requesting job %s", job_id)
+                raise AirflowException(ex)
+            self.log.info("Status of the custom job %s is %s", job.name, job.state.name)
+            if job.state in self.JOB_COMPLETE_STATES:
+                return job
+            self.log.info("Sleeping for %s seconds.", poll_interval)
+            await asyncio.sleep(poll_interval)
+
+    async def wait_for_training_pipeline(
```
Comment (Lee-W) [inline airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py:3506]: I feel the structure of this method is similar to wait_for_custom_job. Not sure whether we can extract the common part of them. Not a blocker of this PR though

File: airflow/providers/google/cloud/operators/vertex_ai/custom_job.py [hunk 5]
```diff
@@ -539,6 +564,94 @@ def on_kill(self) -> None:
...
+            validation_fraction_split=self.validation_fraction_split,
+            test_fraction_split=self.test_fraction_split,
+            training_filter_split=self.training_filter_split,
+            validation_filter_split=self.validation_filter_split,
+            test_filter_split=self.test_filter_split,
+            predefined_split_column_name=self.predefined_split_column_name,
+            timestamp_split_column_name=self.timestamp_split_column_name,
+            tensorboard=self.tensorboard,
+        )
+        custom_container_training_job_obj.wait_for_resource_creation()
```
Comment (Lee-W) [inline airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: Is it possible to do it in triggerer asynchronously?

File: airflow/providers/google/cloud/operators/vertex_ai/custom_job.py [hunk 6]
```diff
@@ -920,9 +1042,98 @@ def on_kill(self) -> None:
...
+            validation_fraction_split=self.validation_fraction_split,
+            test_fraction_split=self.test_fraction_split,
+            training_filter_split=self.training_filter_split,
+            validation_filter_split=self.validation_filter_split,
+            test_filter_split=self.test_filter_split,
+            predefined_split_column_name=self.predefined_split_column_name,
+            timestamp_split_column_name=self.timestamp_split_column_name,
+            tensorboard=self.tensorboard,
+        )
+        custom_python_training_job_obj.wait_for_resource_creation()
```
Comment (Lee-W) [inline airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:1110]: same here

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py (hunk 3 above)
Comment (e-galan) [reply airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py]: I think I just copied it from other methods, so I'm not sure if there was a convention. I will follow your suggestion however.

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py (hunk 4 above)
Comment (e-galan) [reply airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py:3506]: Indeed, the structure is deliberately similar. I tried to follow the same pattern for Vertex AI trigger and hook methods. 
The only difference is that ` wait_for_custom_job()` and `wait_for_training_pipeline()` use different clients and client methods.

File: airflow/providers/google/cloud/operators/vertex_ai/custom_job.py (hunk 5 above)
Comment (e-galan) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: This call normally takes fractions of a second.  I put it here to make sure that all underlying resources are created before querying the name attribute of the Custom Container Training Job object, which corresponds to the training pipeline ID. I needed this pipeline ID for generating a link to the training pipeline **before** deferring the task.

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py (hunk 1 above)
Comment (e-galan) [reply airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py]: Thanks for noticing that! I changed it to a tuple as in the rest of the hook methods that use the `Sequence` type parameters.

File: airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py (hunk 2 above)
Comment (e-galan) [reply airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py:1807]: I guess we could, I also don't find it very readable.

However I just copied these type hints from the existing hook methods, where I think it was introduced deliberately as part of applying PEP-563. If I create an alias I will have to replace this line everywhere.

**Review by potiuk — APPROVED** [review]

File: airflow/providers/google/cloud/operators/vertex_ai/custom_job.py (hunk 5 above)
Comment (potiuk) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: It does impact the event loop - so while it is ok to have it - maybe we could turn it into asyncio-compatible? I generally approve that one, but maybe extracting it into asyncio call would be a good idea?

Comment (e-galan) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: @potiuk I am not sure I understand what event loop is impacted. I'll try to explain my thinking below.

This `wait_for_resource_creation()` method belongs to the `CustomContainerTrainingJob` client class that is used for creating and launching a Vertex AI training pipeline. I call the method because it is recommended to do so  [here](https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform/base.py#L1376-L1405) as the next step after launching a training pipeline with parameter `sync=False`.  Despite its name, the method is not asynchronous and is not part of any event loops AFAIK. 

On the other hand, the existing trigger event loop uses another client class -  `PipelineServiceAsyncClient` to query the training pipeline status until its completion. It is not tied to `wait_for_resource_creation()` directly. This process can take a long time and that is why I chose to defer at this point.

To sum up:
- We launch a training pipeline synchronously by calling CustomContainerTrainingJob's `run()` and   `wait_for_resource_creation()` methods. This whole operation does not take more than a minute.
- We use CustomContainerTrainingJob's attribute for generating a Vertex AI link for the users.
- Then we defer, freeing up the worker slot, and wait until the training pipeline is completed using PipelineServiceAsyncClient's API. This operation could take hours.

Comment (potiuk) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: Ah. yeah. I just saw more context now, then yes - absolutely no problem with it tin this case

Comment (potiuk) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: @Lee-W ? What do you think?  Left the other comment for you to resolve

Comment (Lee-W) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py:635]: Yep, sounds good!

File: tests/providers/google/cloud/operators/test_vertex_ai.py [hunk 7]
```diff
@@ -88,11 +88,18 @@
...
+    CustomPythonPackageTrainingJobTrigger,
+    CustomTrainingJobTrigger,
+    RunPipelineJobTrigger,
+)
 from airflow.utils import timezone
 
 VERTEX_AI_PATH = "airflow.providers.google.cloud.operators.vertex_ai.{}"
 VERTEX_AI_LINKS_PATH = "airflow.providers.google.cloud.links.vertex_ai.{}"
+VERTEX_AI_TRIGGER_PATH = "airflow.providers.google.cloud.triggers.vertex_ai.{}"
+VERTEX_AI_HOOK_PATH = "airflow.providers.google.cloud.hooks.vertex_ai."
```
Comment (Lee-W) [inline tests/providers/google/cloud/operators/test_vertex_ai.py]: May I know whether is it used?

**Review by Lee-W — APPROVED** [review]
Left a few nitpicks. But overall looks good to me!

File: tests/providers/google/cloud/operators/test_vertex_ai.py (hunk 7 above)
Comment (e-galan) [reply tests/providers/google/cloud/operators/test_vertex_ai.py]: Good eyes, thanks!
